File: Tests/CorePromise/02_ErrorUnhandlerTests.m
@import PromiseKit;
@import XCTest;
#import "Infrastructure.h"

@interface WTFError : NSError @end @implementation WTFError
@end


@interface ErrorUnhandlerTests: XCTestCase @end @implementation ErrorUnhandlerTests

- (void)test_68_unhandled_error_handler {
    @autoreleasepool {
        XCTestExpectation *ex = [self expectationWithDescription:@""];

        Injected.errorUnhandler = ^(NSError *error){
            XCTAssertEqual(error.code, 5);
            XCTAssertEqualObjects(@"a", error.domain);
            [ex fulfill];
        };

        (void) [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
            resolve([NSError errorWithDomain:@"a" code:5 userInfo:@{@"1": @2}]);
        }];
    }
    [self waitForExpectationsWithTimeout:2 handler:nil];
}


- (void)test_69_unhandled_handled_returned {
    @autoreleasepool {
        XCTestExpectation *ex1 = [self expectationWithDescription:@"unhandler"];
        XCTestExpectation *ex2 = [self expectationWithDescription:@"initial catch"];

        Injected.errorUnhandler = ^(NSError *error){
            XCTAssertEqual(5, error.code);
            [ex1 fulfill];
        };

        [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
            resolve([NSError errorWithDomain:@"a" code:5 userInfo:nil]);
        }].catch(^(id e){
            [ex2 fulfill];
            return e;
        }).then(^{
            XCTFail();
        });
    }
    [self waitForExpectationsWithTimeout:2 handler:nil];
}

- (void)test_70_unhandled_error_handler_not_called {
    Injected.errorUnhandler = ^(id err){
        XCTFail();
    };

    @autoreleasepool {
        XCTestExpectation *ex1 = [self expectationWithDescription:@""];

        [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
            resolve([NSError errorWithDomain:@"a" code:5 userInfo:nil]);
        }].catch(^{
            return dispatch_promise(^{
                return dispatch_promise(^{
                    @throw @"5";
                });
            });
        }).catch(^{
            [ex1 fulfill];
        });
    }

    [self waitForExpectationsWithTimeout:2 handler:nil];
}

- (void)test_79_unhandled_error_handler_not_called_reject_passed_through {
    @autoreleasepool {
        XCTestExpectation *ex1 = [self expectationWithDescription:@""];

        Injected.errorUnhandler = ^(id e){
            XCTFail();
        };

        [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
            dispatch_promise(^{
                @throw @"1";
            }).catch(resolve);
        }].catch(^{
            [ex1 fulfill];
        });
    }

    [self waitForExpectationsWithTimeout:2 handler:nil];
}

- (void)test_80_unhandled_error_handler_called_if_reject_passed_through {
    @autoreleasepool {
        XCTestExpectation *ex1 = [self expectationWithDescription:@""];
        XCTestExpectation *ex2 = [self expectationWithDescription:@""];

        __block BOOL ex1Fulfilled = NO;

        Injected.errorUnhandler = ^(id e){
            XCTAssert(ex1Fulfilled);
            [ex2 fulfill];
        };

        [AnyPromise promiseWithResolverBlock:^(PMKResolver resolve) {
            dispatch_promise(^{
                @throw @"1";
            }).catch(resolve);
        }].always(^{
            [ex1 fulfill];
            ex1Fulfilled = YES;
        });
    }
    [self waitForExpectationsWithTimeout:2 handler:nil];
}

- (void)test_999_allow_error_subclasses {
    XCTestExpectation *ex1 = [self expectationWithDescription:@""];

    PMKAfter(0.02).then(^{
        return [WTFError errorWithDomain:@"WTF" code:0 userInfo:nil];
    }).catch(^(NSError *e){
        XCTAssertEqualObjects(e.class, WTFError.class);
        [ex1 fulfill];
    });

    [self waitForExpectationsWithTimeout:1 handler:nil];
}

@end

File: Tests/CorePromise/02_ErrorUnhandlerTests.swift
import PromiseKit
import XCTest

class ErrorUnhandlerTests: XCTestCase {

    func testHandlerIsCalledIfErrorIsNotHandled() {
        twice { promise, ex in
            InjectedErrorUnhandler = { err in
                ex.fulfill()
            }
        }
    }

    func testCanOnlyCallSetUnhandlerOnce() {
        // we call PMKSetUnhandledErrorHandler initially in setUp()
        PMKSetUnhandledErrorHandler({ _ in XCTFail() })
        testHandlerIsCalledIfErrorIsNotHandled()
    }

    func testHandlerIsNotCalledIfErrorIsCaught() {
        twice { promise, ex in
            InjectedErrorUnhandler = { err in
                XCTFail()
            }
            promise.catch { error in
                ex.fulfill()
            }
        }
    }

    func testHandlerIsNotCalledIfErrorIsRecovered() {
        twice { promise, ex in
            InjectedErrorUnhandler = { err in
                XCTFail()
            }
            _ = promise.recover { error -> Promise<Int> in
                return Promise(value: 1)
            }.always {
                ex.fulfill()
            }
        }
    }

    func testHandlerIsStillCalledIfErrorIsNotRecovered() {
        twice { promise, ex in
            InjectedErrorUnhandler = { err in
                ex.fulfill()
            }
            _ = promise.recover { error -> Int in
                throw error
            }
        }
    }

    func testHandlerIsCalledOnceIfRecoveryFailsWithADifferentError() {
        twice { promise, ex in
            InjectedErrorUnhandler = { err in
                ex.fulfill()
            }
            promise.recover { error -> Promise<Int> in
                firstly { throw Error.dummy }
            }
        }
    }

    func testHandlerIsNotCalledIfErrorRecoveryFailsWithSameErrorButEventuallyErrorIsCaught() {
        twice { promise, ex2 in
            let ex1 = expectation(description: "")

            InjectedErrorUnhandler = { err in
                XCTFail()
            }
            promise.recover { error -> Int in
                throw error
            }.then { x in
                XCTFail()
            }.always {
                ex1.fulfill()
            }.catch { err in
                ex2.fulfill()
            }
        }
    }

    func testDoubleRejectDoesNotTriggerUnhandler() {
        enum Error: Swift.Error {
            case test
        }

        InjectedErrorUnhandler = { err in
            XCTFail()
        }

        let (p, _, r) = Promise<Void>.pending()

        let ex1 = expectation(description: "")
        let ex2 = expectation(description: "")
        let ex3 = expectation(description: "")
        let ex4 = expectation(description: "")

        after(interval: .milliseconds(100)).then { _ -> Void in r(Error.test); ex1.fulfill() }
        after(interval: .milliseconds(150)).then { _ -> Void in r(Error.test); ex2.fulfill() }.always(execute: ex3.fulfill)

        p.catch { error in
            ex4.fulfill()
        }

        waitForExpectations(timeout: 1, handler: nil)
    }

    func testPassThrough() {
        let ex = expectation(description: "")

        InjectedErrorUnhandler = { err in
            ex.fulfill()
        }

        enum Error: Swift.Error {
            case test
        }

        Promise<Void> { _, reject in
            after(interval: .milliseconds(100)).then {
                throw Error.test
            }.catch(execute: reject)
        }

        waitForExpectations(timeout: 1, handler: nil)
    }

    func testConsumedPromiseStaysConsumedAsAnyPromise() {
        enum Error: Swift.Error {
            case test
        }

        InjectedErrorUnhandler = { err in
            XCTFail()
        }

        let ex1 = expectation(description: "")

        let p: Promise<Int> = firstly {
            throw Error.test
        }

        XCTAssertTrue(p.isRejected)

        let anyp = AnyPromise(p)

        p.catch { err in
            ex1.fulfill()
        }

        waitForExpectations(timeout: 1)

        print(anyp)
    }


//MARK: helpers

    private func twice(body: (Promise<Int>, XCTestExpectation) -> Void) {
        autoreleasepool {
            let ex = expectation(description: "Sealed")
            body(Promise<Int>(error: Error.dummy), ex)
        }
        waitForExpectations(timeout: 1)

        autoreleasepool {
            let ex = expectation(description: "Unsealed")
            let p = Promise { fulfill, _ in
                fulfill(1)
            }.then { (_: Int) -> Int in
                throw Error.dummy
            }
            body(p, ex)
        }
        waitForExpectations(timeout: 1)
    }

}

private enum Error: Swift.Error {
    case dummy
}
